transient/configuration.py

```diff
@@ -0,0 +1,331 @@
+
+    Example:
+
+        >>> args = Config({'arg1': 1, 'arg2': 2})
+        >>> args['arg1'] == args.arg1
+
+    """
+
+    def __init__(self, *args: List[Dict[Any, Any]], **kwargs: Dict[Any, Any]) -> None:
+        super(Config, self).__init__(*args, **kwargs)
+        for arg in args:
+            if isinstance(arg, dict):
+                for k, v in arg.items():
+                    self[k] = v
+
+        if kwargs:
+            for k, v in kwargs.items():
+                self[k] = v
+
+    def __getattr__(self, attr: Any) -> Any:
+        return self.get(attr)
+
+    def __setattr__(self, key: Any, value: Any) -> None:
+        self.__setitem__(key, value)
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        super(Config, self).__setitem__(key, value)
+        self.__dict__.update({key: value})
+
+    def __delattr__(self, item: Any) -> None:
+        self.__delitem__(item)
+
+    def __delitem__(self, key: Any) -> None:
+        super(Config, self).__delitem__(key)
+        del self.__dict__[key]
+
+
+class _TransientConfigSchema(Schema):
+    image = fields.List(fields.Str(), missing=[])
+    """Defines a common schema for the Transient configurations and validates
+       the fields during deserialization
+    """
+    image_backend = fields.Str(allow_none=True)
+    image_frontend = fields.Str(allow_none=True)
+    name = fields.Str(allow_none=True)
+
+    # marshmallow's decorator pre_load() is untyped, forcing
+    # remove_unset_options() to be untyped. Therefore, we ignore it to
+    # silence the type checker
+    @pre_load  # type: ignore
+    def remove_unset_options(
+        self, config: Dict[Any, Any], **kwargs: Dict[Any, Any]
+    ) -> Dict[Any, Any]:
+        """Removes any option that was not set in the command line
+        """
+        processed_config = {}
+        for option, value in config.items():
+            if _option_was_set_in_cli(value):
+                processed_config[option] = value
+
+        return processed_config
+
+    # marshmallow's decorator post_load() is untyped, forcing create_args()
+    # to be untyped. Therefore, we ignore it to silence the type checker
+    @post_load  # type: ignore
+    def create_config(self, data: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> Config:
+        """Returns the Config dictionary after a schema is loaded and validated
+        """
+        return Config(**data)
+
+
+class _TransientDeleteConfigSchema(_TransientConfigSchema):
+    """Defines the schema for the Transient-delete configuration and validates
+       the fields during deserialization
+    """
+
+    force = fields.Bool(missing=False)
+
+
+class _TransientListConfigSchema(_TransientConfigSchema):
+    """Defines the schema for the Transient-list configuration and validates
+       the fields during deserialization
+
+       Note that this class is a wrapper to maintain symmetry with the other
+       schemas.
+    """
+
+    pass
```

Comment:
I'm not sure that allowing the config stuff for list/delete is wise. I would at least like to see some tests for it. The issue I think of here is that there are slightly different meanings to the usage of `-name` and `-image` between deleteing and running a VM. For example:

Suppose you have a config that specifies a image but no name, so it creates a VM with a UUID name:

```
[transient]
image = ["centos/7:2004.01"]
```

Then you run `transient delete -c /the/above/config`, you will actually delete the backend image "centos/7:2004.01", because that is the behavior of `transient delete -image centos/7:2004.01` (deleting an image with no name deletes the backend image and all associated frontend images that were backed by that image).

That said, it could just be considered a user error (and they will still get the warning), so I'm probably Ok with this, it should just have tests for list/delete with a config file.

Reply:
I agree that using a configuration file for list or delete would be a bad idea. I think it would be too easy to accidentally execute `transient delete` with a given config file, causing a negative user experience.

That's why I did not implement that functionality. Using a configuration file is only possible for `Transient-run`. The reason for multiple schema classes is to have a symmetric configuration schemas between `run`, `list`, and `delete`.

Essentially,  all command line arguments are passed to marshmallow for initialization (setting default values if required) and validation. Then, marshmallow returns the validated Config object for use. This happens for `run`, `list`, and `delete`.